src/main/cpp/src/utils.hpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+
+namespace spark_rapids_jni {
+
+/**
+ * Represents timestamp with microsecond accuracy.
+ * Max year is six-digits, approximately [-300000, 300000]
+ */
+struct ts_segments {
+  __device__ ts_segments()
+    : year(1970), month(1), day(1), hour(0), minute(0), second(0), microseconds(0)
+  {
+  }
+
+  /**
+   * Is the segments of the timestamp valid?
+   */
+  __device__ bool is_valid_ts() const { return is_valid_date() && is_valid_time(); }
+
+  /**
+   * Get epoch day. Can handle years in the range [-1,000,000, 1,000,000].
+   * Refer to https://howardhinnant.github.io/date_algorithms.html#days_from_civil
+   * Spark year is approximately in range [-300,000, 300,000]
+   * std::chrono::year range is [-32,767 , 32,767], here can not use std::chrono::year_month_day
```

Comment:
If this is true, does that mean any time we use a year outside of this range it is highly probably that std::chrono and hence CUDF will return the wrong result? We often limit our tests to years within a valid python range, which is smaller than both of these.

Reply:
Yes that is true. If you look into cudf code such as `convert_datetime.cu`, `cuda::std::chrono::year` is used in many places and should have the same restriction.

Reply:
Due to such issue, I don't think we should pursue supporting large year numbers as Spark does - large years may not be used in the real world. Otherwise, it would be not a simple task to fix all the related code.

Reply:
For this casting string to timestamp PR, it supports large years. We can file a follow up issue to track other related code.